feat(plugin): Add ability to push multiple docker images

DanielHabenicht opened the pull request:
This feature allows specifying the plugin multiple times in order to push multiple images to a
docker repository.

#4

DanielHabenicht commented:
I don't know why but the tests in the 
https://github.com/DanielHabenicht/semantic-release-docker/blob/feat/multiple-images/src/shared-logic.spec.ts
file are not executed.

cmurczek commented:
Hi @DanielHabenicht, with this PR you'd basically revert my changes to mock/stub the docker dependency during unit tests. I've added more test as examples for how I'd like to proceed. Please review these and update your PR.

DanielHabenicht commented:
Ahh, ok. I think I misunderstood. I will include your changes as units tests and add my tests as e2e Tests.

cmurczek commented:
I actually meant for you to refactor your tests to use sinon for mocking/stubbing, so that testruns are no longer depending on docker being available on build/test agents. An added benefit is that testruns take a LOT less time, eg currently down to 85ms per test run opposed to 1 – 3s per testcase.

Sent from my Windows 10 device

________________________________
From: DanielHabenicht <notifications@github.com>
Sent: Wednesday, May 1, 2019 12:53:16 PM
To: iteratec/semantic-release-docker
Cc: Christoph Murczek; Comment
Subject: Re: [iteratec/semantic-release-docker] feat(plugin): Add ability to push multiple docker images (#7)


Ahh, ok. I think I misunderstood. I will include your changes as units tests and add my tests as e2e Tests.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub<https://github.com/iteratec/semantic-release-docker/pull/7#issuecomment-488255865>, or mute the thread<https://github.com/notifications/unsubscribe-auth/AGDFOJM2DR27PW535IG7CGTPTFZBZANCNFSM4HI2VTQQ>.

DanielHabenicht commented:
Finally, I had some time to refactor the code. 
I've used the sinon mocks and stubs in the spec files and made it optional to run the tests that are docker dependent. 

I take your point on running the tests a lot faster for development, that's why I refactored it. 
But I think it is also important to test the whole plugin (including the dockerode part), that's why I included them as e2e tests that can be run separately (this way we can test if the plugin can interact with a real docker API, thus test if the current version of dockerode does work properly in our use case or even test if the publish step works later on)

Have a look and feel free to comment. 
I will refactor the other PR after this gets merged.

cmurczek commented on README.md:
````diff
@@ -84,4 +64,95 @@ minimum configuration:
+publish(config, context);
+```
+
+2.  Simply run the "Debug" VS Code Task
````
Why would the debugger act on the develop.ts file described above? There is no reference to it anywhere. Or did you forget to commit changes to you launch.json?
I'd recommend to rather use the debug feature of the mocha vscode extension, since it supports a TDD approach to debug a specific scenario described by a test case.

DanielHabenicht replied:
I did forget to commit it indeed. I like the TDD approach as well but sometimes it is way easier to be in control of everything and just debug the file/function you want.

cmurczek commented on .gitignore:
```diff
@@ -10,3 +10,5 @@ dist
 
 *.log
+
+develop.ts
```
please remove the develop.ts file and it's mentioning here.

DanielHabenicht replied:
as written in the other comment this is for the developers safeguard

cmurczek commented on Dockerfile:
```diff
@@ -0,0 +1,2 @@
+FROM scratch
+ARG TEST
```
Don't burden yourself with creating actual images while developing this plugin. There is no benefit over (correctly) mocking/stubbing dockerode.

DanielHabenicht commented:
>I do get your motivation for bringing in e2e tests, but this way they provide no benefit over good unit tests. They do however cost time, especially when they are part of every build.


There are numerous reasons and benefits for the e2e tests: 
 1. The `e2e` tests actually check on the machine if the plugin has done what it said it did. 
With the Unit-Tests only the tools 'said' output is tested. For Example the Bug I fixed before was not detected by your unit tests, because it did only check the tool's output, not the actual behaviour of the tool. That's also the reason why I wrote them in the first place. 
2. The `e2e` tests are independent of the tools you use (ideally). 
You might want to replace dockerode in the future because it gets deprecated or update to the next version. This does change the API of the internally used tools and you have to rewrite your unit tests. While rewriting who checks that the tests and changes you do are correct? (The e2e tests, because they only check the outcome)

If you think the build time is too long I can make them run in parallel. But I think that waiting 22 seconds are worth the amount of time needed for others to test, verify, file a bug and resolve the issue coming from not testing it. 
>If we added e2e tests they should be truly e2e, meaning actually using semantic-release with this plugin to build a docker image. However, for now, at least I don't see the need for that.

I think of these as integration tests, that's also a good idea but I do align with your view.

>I do like the original idea of the PR though. I'm not sure if the way it's implemented is the best strategy. What's your take on something like
{ plugins: [ prepare: [{ path: "@iteratec/semantic-release-docker", images: [{ additionalTags: [], imageName: "", registryUrl: "", repositoryName: "", }] }] }
To me it feels more natural to configure the plugin once, and have it deal with multiple images, instead of loading multiple instances of the same plugin...

I think the way it is implemented is the way to go, with the option to specify multiple images you are bringing in another way of using the plugin, while not fixing the way it already can be used. 
For reference the `@semantic-release/exec` command does it the same way. But it might be a good addition in the future to provide both ways.

DanielHabenicht commented on src/verifyConditions/index.e2e.ts:
```diff
@@ -0,0 +1,172 @@
+      return expect(verifyConditions(config, context)).to.not.eventually.be.rejectedWith(
+        'Environment variable DOCKER_REGISTRY_PASSWORD must be set in order to login to the registry.',
+      );
+    });
```
done

DanielHabenicht commented on src/verifyConditions/index.e2e.ts:
```diff
@@ -0,0 +1,172 @@
+      );
+    });
+
+    it('should default to docker hub if no registry is specified', async function() {
```
done

DanielHabenicht commented on src/verifyConditions/index.e2e.ts:
```diff
@@ -0,0 +1,172 @@
+
+    it('should default to docker hub if no registry is specified', async function() {
+      this.timeout(10000);
+      await buildImage(imageName);
```
now superfluous, as the test is removed

DanielHabenicht commented on src/verifyConditions/index.e2e.ts:
```diff
@@ -0,0 +1,172 @@
+      );
+    });
+
+    it('should throw if no imagename is provided', function() {
```
done

DanielHabenicht commented on README.md:
````diff
@@ -84,4 +64,95 @@ minimum configuration:
+verifyConditions(config, context);
+prepare(config, context);
+publish(config, context);
+```
````
In order to safeguard the user from accidentally committing their credentials. That`s also the reason why `develop.ts` is in the `.gitignore`.

DanielHabenicht commented on src/prepare/index.e2e.ts:
```diff
@@ -0,0 +1,230 @@
+      expect(imagelist2.length).to.equal(1);
+    });
+
+    it('should add multiple images', async function() {
```
there already is one

DanielHabenicht commented on src/prepare/index.ts:
```diff
@@ -1,54 +1,57 @@
+    preparePlugins.map((preparePlugin) => {
+      const docker = dockerode ? dockerode : new Dockerode();
+      const image = docker.getImage(preparePlugin.imageName);
+      const tags = getImageTagsFromConfig(preparePlugin, context);
```
Why?

DanielHabenicht commented on src/prepare/index.ts:
```diff
@@ -1,54 +1,57 @@
+      )
+        .then((data) => {
+          if (!prepared) {
+            prepared = true;
```
good catch! Done.

DanielHabenicht commented on src/publish/index.ts:
```diff
@@ -15,42 +16,64 @@ export interface PublishedRelease {
+
+  return Promise.all(
+    preparePlugins.map(preparePlugin => {
+      const tags = getImageTagsFromConfig(preparePlugin, context);
```
Why should we duplicate code two times?

DanielHabenicht commented on src/publish/index.ts:
```diff
@@ -15,42 +16,64 @@ export interface PublishedRelease {
+      const image = docker.getImage(imageName);
+      const options: PushOptions = {
+        password: process.env.DOCKER_REGISTRY_PASSWORD!,
+        serveraddress: getRegistryUrlFromConfig(preparePlugin),
```
Why should we duplicate code two times?

DanielHabenicht commented on src/shared-logic.ts:
```diff
@@ -0,0 +1,48 @@
+  );
+}
+
+export function getRegistryUrlFromConfig(config: DockerPluginConfig): string {
```
But why? The Function name provides the insight of what it is doing + the function is used two times.

DanielHabenicht commented on src/shared-logic.ts:
```diff
@@ -0,0 +1,48 @@
+    : '';
+}
+
+export function getImageTagsFromConfig(config: DockerPluginConfig, context: SemanticReleaseContext): string[] {
```
But why? The Function name provides the insight of what it is doing + the function is used two times.

DanielHabenicht commented on src/shared-logic.ts:
```diff
@@ -0,0 +1,48 @@
+ * Get Authentication object from Environment Variables
+ * Throws Error if Variables are not set.
+ */
+export function getCredentials(): Credentials {
```
sry, I forgot to refactor the usage in the publish step, now the function is used twice.

DanielHabenicht commented on src/verifyConditions/index.spec.ts:
```diff
@@ -89,29 +194,12 @@ describe('@iteratec/semantic-release-docker', function() {
-      expect(dockerStub.checkAuth.firstCall.args[0]).to.deep.equal(expected);
-    });
-
-    it('should default to docker hub if no registry is specified', async function() {
```
Because it basically is a duplicate of "should use the registry from the config" the test checks if the string put into the config is returned. The main difference is that one string is `''` and the other is `'my_private_registry'`

DanielHabenicht commented on src/verifyConditions/index.ts:
```diff
@@ -1,61 +1,74 @@
+    // Check Authentication
+    const auth: Authentication = {
+      ...cred,
+      serveraddress: getRegistryUrlFromConfig(preparePlugin),
```
Why should we duplicate code two times?

DanielHabenicht commented on src/verifyConditions/index.ts:
```diff
@@ -1,61 +1,74 @@
+      serveraddress: getRegistryUrlFromConfig(preparePlugin),
+    };
+
+    return docker
```
Thanks! Using `Promise.all(` now.

DanielHabenicht commented on src/shared-logic.spec.ts:
```diff
@@ -0,0 +1,73 @@
+      expect(constructImageName(config)).to.be.equal('registry/repo/test');
+    });
+
+    it('should use the registry from the config', function() {
```
Why?

DanielHabenicht commented on src/prepare/index.ts:
```diff
@@ -1,54 +1,57 @@
-    tags = tags.concat(preparePlugin.additionalTags);
+  dockerode?: Dockerode,
+): Promise<any[]> {
+  if (!verified) {
```
Yep, I think that's not handled in the way semantic-release imagined it. With every step being independent of another. At the moment every plugin step executes the previous step if it wasn't executed before. (that is in master)

As this is a breaking Change we should either do this in a seperate PR or I can change it in here.

DanielHabenicht commented on src/prepare/index.ts:
```diff
@@ -1,54 +1,57 @@
-        tag: imagetag,
-      });
+    preparePlugins.map((preparePlugin) => {
+      const docker = dockerode ? dockerode : new Dockerode();
```
done

DanielHabenicht commented on README.md:
```diff
@@ -7,26 +7,13 @@
 
 A [semantic-release](https://github.com/semantic-release/semantic-release) plugin to use semantic versioning for docker images.
 
```
I like it when relevant information comes first. In my mind the Question _How do I use the Plugin?_ has to be the first one answered. Not which steps do what. This is something more specific that can be read by scrolling down.

DanielHabenicht commented:
@cmurczek any updates?
